### ISSUE TITLE
Investigate vegalite in a Shiny app

### ISSUE THREAD

**Issue by jhofman** [jhofman/datamations#29]
Related to #27, how feasible is it to render a vegalite plot inside of a Shiny app?

To check this out, let's try a simple Shiny app with a dropdown that lets you choose between a [bar chart](https://vega.github.io/editor/#/examples/vega-lite/bar) and a [colored bar chart](https://vega.github.io/editor/#/examples/vega-lite/bar_color_disabled_scale).

The JSON can be hard-coded in. If successful hopefully we can port to a working datamation version that mirrors something like [this](https://nshrest.shinyapps.io/datawhats/).

**Comment by sharlagelfand** [comment]
Here's a super basic app that lets you choose between the bar and colored bar charts: https://sharlagelfand.shinyapps.io/vegalite/ (and the [code](https://github.com/jhofman/datamations/blob/vegalite-shiny/sandbox/vegalite-shiny-app/))

Everything is pretty hard coded, but I imagine as we go forward (in general, not just with Shiny) will need to make this into an [html widget](https://github.com/ramnathv/htmlwidgets), which is what r2d3, leaflet, vegawidget, etc all use

**Comment by jhofman** [comment]
Awesome!

Can you point @giorgi-ghviniashvili to your favorite "intro to Shiny apps" resource(s)?

**Comment by sharlagelfand** [comment]
Definitely! @giorgi-ghviniashvili, the [Mastering Shiny](https://mastering-shiny.org/) book is really awesome for foundational information, especially on how much (all?) of Shiny is just HTML, CSS, and javascript under the hood :) For tutorial-style, I like RStudio's materials starting [here](https://shiny.rstudio.com/tutorial/written-tutorial/lesson1/)

**Comment by jhofman** [comment]
@sharlagelfand, just making some notes from our chat.

next up here is to try changing to an html widget so we can see how passing data goes. for instance, making the drop-down be a selection between one of two different datasets, and then doing a simple plot based on the dataset that's selected. 

also, you mentioned that the [javascript for r](https://javascript-for-r.com) book might be useful.

**Comment by sharlagelfand** [comment]
I have a working prototype! 🎉 Will see if I can figure out deploying easily (a bit complicated with the structure required by htmlwidgets / having an app embedded in the package) but otherwise will demo quickly on our call tomorrow.

**Comment by jhofman** [comment]
prototype looked great. can you commit and post a link?

next step we discussed was to add a text box that takes tidyverse analysis code as input and produces vegalist specs as json for each step somewhere else on the page. (could simply print them all out for now.) then if we can render those specs in the page after that we'd have a nice working version of what could become the datamation editor as per #27.

also, at some point it would be great to document the "flow" of things going from R to javascript to vega, etc.

**Comment by sharlagelfand** [comment]
Super rough draft, but updated [here](https://sharlagelfand.shinyapps.io/datamations/) is a Shiny app that lets you select a data set, as well as variables to group by, summary variable, and summary function. Figured it's a little safer right now to force the selection of groups/summary variables rather than parse free text input.

It ended up easier to render the plots rather than output the actual JSON specs, but I'll continue iterating on this. Also using the existing grouping of colour -> shape, and plots are cramped, etc etc, but for a proof of concept on the stages spec output and Shiny side!

**Comment by jhofman** [comment]
this is great, very nice!

@giorgi-ghviniashvili, next step would be to incorporate your code to datamate the different frames. we can discuss the next time we chat.

**Comment by sharlagelfand** [comment]
We indeed have the app working, so I'll close this!
